keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
+corresponding the PodQoSResources message in PodSandboxConfig in the CRI
+API.
+
+```diff
+ type PodSpec struct {
+@@ -3062,6 +3069,10 @@ type PodSpec struct {
+        ResourceClaims []PodResourceClaim
++       // Pod-level resources. Claims, requests and limits are not allowed
++       // to be specified for pods.
++       // +optional
++       Resources ResourceRequirements
+ }
+```
+
+There is already an ongoing effort to add [Pod level resource limits][kep-2837]
+that aims at adding a pod level `Resources` field in a similar fashion. Thus,
+we opt for adding ResourceRequirements insted of QoSResources directly into the
+PodSpec.
+
+As an example, a Pod requesting class "fast" of a (exemplary) pod-level QoS
+resource named "network", with one container requesting class "gold" of
+container-level QoS resource named "rdt":
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: qos-resource-example
+spec:
+  resources:
+    qosResources:
+      network: fast
+  containers:
+  - name: cnt
+    image: nginx
+    resources:
+      qosResources:
+        rdt: gold
+```
+
+#### NodeStatus
+
+We extend NodeStatus to list available QoS-class resources on a node, This
+consists of the list of the available QoS-class resource types and the classes
+available within each of these resource types.
+
+
+
+```diff
+ type NodeStatus struct {
+@@ -4444,6 +4482,11 @@ type NodeStatus struct {
+        // Status of the config assigned to the node via the dynamic Kubelet config feature.
+        Config *NodeConfigStatus
++       // QoSResources contains information about the QoS resources that are
++       // available on the node.
++       // +featureGate=QoSResources
++       // +optional
++       QoSResources QoSResourceStatus
+ }
+
+...
+
++// QoSResourceStatus describes QoS resources available on the node.
++type QoSResourceStatus struct {
++       // PodQoSResources contains the QoS resources that are available for pods
++       // to be assigned to.
++       PodQoSResources []QoSResourceInfo
++       // ContainerQoSResources contains the QoS resources that are available for
++       // containers to be assigned to.
++       ContainerQoSResources []QoSResourceInfo
++}
+
++// QoSResourceInfo contains information about one QoS resource type.
++type QoSResourceInfo struct {
++       // Name of the resource.
++       Name QoSResourceName
++       // Mutable is set to true if the resource supports in-place updates.
++       Mutable bool
++       // Classes available for assignment.
++       Classes []QoSResourceClassInfo
++}
+
++// QoSResourceClassInfo contains information about single class of one QoS
++// resource.
++type QoSResourceClassInfo struct {
++       // Name of the class.
++       Name string
++       // Capacity is the number of maximum allowed simultaneous assignments into this class
++       // Zero means "infinite" capacity i.e. the usage is not restricted
++       // +optional
++       Capacity int64
++}
+```
```

Comment:
We discussed this on the call, but capturing here.

Consider adding a `QoSClass` API resource.  It doesn't need to be very rich, just a resource name and class name, maybe.  This lets end-users know that, as a cluster admin you intended for this class to exist.  Also, they can then `kubectl get qosclass` and see what they have available.    Depending on how we answer defaults, maybe this helps there too?

@msau42 brought this up and had another argument, but I failed to capture it in my notes...

Reply:
The scenario I was thinking was also on the "portability" question, but instead of kubernetes defining the semantics of what the resource means with well-known names, what if it can be admin-defined?

For example, what if different runtimes or different hardware vendors have different, but close-enough implementations of "cache-line"? A platform admin who's providing a PaaS to their users wants to define a portable class that could be leveraged on all the hardware platforms they support. So maybe they would want to define their own "cache" class, and then be able to specify some mapping that "cache/gold" can be satisfied by "rdt/gold" or "foo/gold", "cache/silver" maps to "foo/silver", etc.

Reply:
FWIW I am still not sure about this aspect - maybe mark it as UNRESOLVED in the KEP and we can revisit as we get more clarity?

Reply:
@thockin can you confirm what you actually want to be marked as unresolved? I hastily read your comment and marked the Capacity field in QOSResourceClassInfo as unresolved but I don't think that's what you wanted. If I'm not mistaken you want to add a section about a separate QOSClass (and QOSResource?) K8s API objects and mark that as unresolved(?)

Reply:
ping @thockin ^